File: go/internal/e2e/llm_inference_handler_e2e_test.go
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

package e2e

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"net/url"
	"strings"
	"sync/atomic"
	"testing"

	"github.com/coder/websocket"
	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
)

const (
	llmHandlerHTTPText = "OK from synthetic HTTP upstream."
	llmHandlerWSText   = "OK from synthetic WS upstream."
)

type llmHandlerCounters struct {
	httpRequests       atomic.Int32
	httpResponses      atomic.Int32
	wsRequestMessages  atomic.Int32
	wsResponseMessages atomic.Int32
	upstreamWSRequests atomic.Int32
}

func llmSSEBody(text, respID string) string {
	var sb strings.Builder
	for _, event := range llmResponsesEvents(text, respID) {
		sb.WriteString(llmSSE(event["type"].(string), event))
	}
	return sb.String()
}

// startFakeUpstream brings up a real HTTP upstream (catalog / policy /
// responses-SSE) and a real WebSocket upstream that echoes the ordered
// /responses events per inbound message.
func startFakeUpstream(t *testing.T, counters *llmHandlerCounters) (httpURL, wsURL string) {
	t.Helper()

	httpSrv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		path := strings.ToLower(strings.SplitN(r.URL.Path, "?", 2)[0])
		_ = r.Body.Close
		switch {
		case strings.HasSuffix(path, "/models"):
			w.Header().Set("content-type", "application/json")
			_, _ = w.Write([]byte(llmModelCatalog(llmWSSupportedEndpoints)))
		case strings.HasSuffix(path, "/models/session"):
			w.Header().Set("content-type", "application/json")
			_, _ = w.Write([]byte("{}"))
		case strings.Contains(path, "/policy"):
			w.Header().Set("content-type", "application/json")
			_, _ = w.Write([]byte(`{"state":"enabled"}`))
		case strings.HasSuffix(path, "/responses"):
			w.Header().Set("content-type", "text/event-stream")
			_, _ = w.Write([]byte(llmSSEBody(llmHandlerHTTPText, "resp_stub_http")))
		default:
			w.Header().Set("content-type", "application/json")
			w.WriteHeader(http.StatusNotFound)
			_, _ = w.Write([]byte(`{"error":"not_found"}`))
		}
	}))
	t.Cleanup(httpSrv.Close)

	wsSrv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		c, err := websocket.Accept(w, r, &websocket.AcceptOptions{InsecureSkipVerify: true})
		if err != nil {
			return
		}
		defer c.Close(websocket.StatusNormalClosure, "")
		c.SetReadLimit(-1)
		bg := context.Background()
		for {
			_, _, readErr := c.Read(bg)
			if readErr != nil {
				return
			}
			counters.upstreamWSRequests.Add(1)
			for _, event := range llmResponsesEvents(llmHandlerWSText, "resp_stub_ws") {
				raw, _ := json.Marshal(event)
				if err := c.Write(bg, websocket.MessageText, raw); err != nil {
					return
				}
			}
		}
	}))
	t.Cleanup(wsSrv.Close)

	return httpSrv.URL, "ws://" + strings.TrimPrefix(wsSrv.URL, "http://")
}

type llmRewritingRoundTripper struct {
	base     *url.URL
	counters *llmHandlerCounters
	inner    http.RoundTripper
}

func (rt *llmRewritingRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
	rt.counters.httpRequests.Add(1)
	req.URL.Scheme = rt.base.Scheme
	req.URL.Host = rt.base.Host
	req.Host = rt.base.Host
	req.Header.Set("x-test-mutated", "1")
	resp, err := rt.inner.RoundTrip(req)
	if err != nil {
		return nil, err
	}
	rt.counters.httpResponses.Add(1)
	resp.Header.Set("x-test-response-mutated", "1")
	return resp, nil
}

func TestLlmInferenceHandler(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	counters := &llmHandlerCounters{}
	httpURL, wsURL := startFakeUpstream(t, counters)

	httpBase, err := url.Parse(httpURL)
	if err != nil {
		t.Fatalf("Failed to parse upstream URL: %v", err)
	}
	wsBase, err := url.Parse(wsURL)
	if err != nil {
		t.Fatalf("Failed to parse upstream ws URL: %v", err)
	}

	handler := &copilot.LlmRequestHandler{
		Transport: &llmRewritingRoundTripper{
			base:     httpBase,
			counters: counters,
			inner:    http.DefaultTransport.(*http.Transport).Clone(),
		},
		OpenWebSocket: func(rctx *copilot.LlmRequestContext) (copilot.CopilotWebSocketHandler, error) {
			parsed, perr := url.Parse(rctx.URL)
			if perr != nil {
				return nil, perr
			}
			parsed.Scheme = wsBase.Scheme
			parsed.Host = wsBase.Host
			fwd := copilot.NewForwardingWebSocketHandler(parsed.String(), rctx.Headers)
			fwd.OnSendRequestMessage = func(data []byte) []byte {
				counters.wsRequestMessages.Add(1)
				return data
			}
			fwd.OnSendResponseMessage = func(data []byte) []byte {
				counters.wsResponseMessages.Add(1)
				return data
			}
			return fwd, nil
		},
	}

	client := newLlmClient(ctx, handler, "COPILOT_EXP_COPILOT_CLI_WEBSOCKET_RESPONSES=true")
	t.Cleanup(func() { client.ForceStop() })

	if err := client.Start(t.Context()); err != nil {
		t.Fatalf("Failed to start client: %v", err)
	}

	session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
		OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
	})
	if err != nil {
		t.Fatalf("Failed to create session: %v", err)
	}

	result, err := session.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: "Say OK."})
	if err != nil {
		t.Fatalf("send_and_wait failed: %v", err)
	}
	_ = session.Disconnect()

	// The HTTP seam fired — the runtime issued model-layer GETs (catalog,
	// policy) and possibly a single-shot inference through the RoundTripper.
	if counters.httpRequests.Load() == 0 {
		t.Fatal("Expected the HTTP RoundTripper to fire")
	}
	if counters.httpResponses.Load() == 0 {
		t.Fatal("Expected the HTTP response mutation to fire")
	}

	// The WebSocket seam fired — the main agent turn went over the WS path and
	// we observed messages in both directions.
	if counters.wsRequestMessages.Load() == 0 {
		t.Fatal("Expected runtime → upstream ws messages")
	}
	if counters.wsResponseMessages.Load() == 0 {
		t.Fatal("Expected upstream → runtime ws messages")
	}
	if counters.upstreamWSRequests.Load() == 0 {
		t.Fatal("Expected the upstream WS to receive request messages")
	}

	// Validate the final assistant response arrived (guards against truncated captures)
	text := assistantText(result)
	if !strings.Contains(text, "OK from synthetic") || !strings.Contains(text, "upstream") {
		t.Fatalf("Expected synthetic upstream content in assistant reply, got %q", text)
	}
}

File: java/src/test/java/com/github/copilot/LlmInferenceHandlerE2ETest.java
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

package com.github.copilot;

import static com.github.copilot.LlmInferenceTestSupport.assistantText;
import static com.github.copilot.LlmInferenceTestSupport.newLlmClient;
import static com.github.copilot.LlmInferenceTestSupport.setupCapiAuth;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.InputStream;
import java.net.URI;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import com.github.copilot.generated.AssistantMessageEvent;
import com.github.copilot.rpc.MessageOptions;
import com.github.copilot.rpc.PermissionHandler;
import com.github.copilot.rpc.SessionConfig;

/**
 * Verifies that the runtime's model-layer traffic can be forwarded through the
 * idiomatic {@link LlmRequestHandler} seams to a real upstream: an HTTP send
 * override that mutates the request/response and a forwarding
 * {@link CopilotWebSocketHandler} that observes messages in both directions.
 */
public class LlmInferenceHandlerE2ETest {

    private static E2ETestContext ctx;

    @BeforeAll
    static void setup() throws Exception {
        ctx = E2ETestContext.create();
    }

    @AfterAll
    static void teardown() throws Exception {
        if (ctx != null) {
            ctx.close();
        }
    }

    private static String rewriteHost(String base, URI original) {
        String path = original.getRawPath() == null ? "" : original.getRawPath();
        String query = original.getRawQuery();
        return base + path + (query != null ? "?" + query : "");
    }

    @Test
    void forwardsThroughIdiomaticHandler() throws Exception {
        setupCapiAuth(ctx);

        AtomicInteger httpRequests = new AtomicInteger();
        AtomicInteger httpResponses = new AtomicInteger();
        AtomicInteger wsRequestMessages = new AtomicInteger();
        AtomicInteger wsResponseMessages = new AtomicInteger();

        try (FakeUpstreamServer upstream = new FakeUpstreamServer("OK from synthetic HTTP upstream.",
                "OK from synthetic WS upstream.")) {

            String httpBase = upstream.httpUrl();
            String wsBase = upstream.wsUrl();

            LlmRequestHandler handler = new LlmRequestHandler() {
                @Override
                protected HttpResponse<InputStream> sendHttp(HttpRequest request, LlmRequestContext rctx)
                        throws Exception {
                    httpRequests.incrementAndGet();
                    URI rewritten = URI.create(rewriteHost(httpBase, request.uri()));
                    HttpRequest.Builder builder = HttpRequest.newBuilder().uri(rewritten);
                    request.bodyPublisher().ifPresentOrElse(bp -> builder.method(request.method(), bp),
                            () -> builder.method(request.method(), HttpRequest.BodyPublishers.noBody()));
                    request.headers().map().forEach((name, values) -> {
                        for (String value : values) {
                            try {
                                builder.header(name, value);
                            } catch (IllegalArgumentException ignored) {
                                // Restricted header rejected by java.net.http; skip it.
                            }
                        }
                    });
                    builder.header("x-test-mutated", "1");
                    HttpResponse<InputStream> response = httpClient()
                            .sendAsync(builder.build(), HttpResponse.BodyHandlers.ofInputStream()).get();
                    httpResponses.incrementAndGet();
                    return response;
                }

                @Override
                protected CopilotWebSocketHandler openWebSocket(LlmRequestContext rctx) {
                    String rewritten = rewriteHost(wsBase, URI.create(rctx.url()));
                    return new ForwardingWebSocketHandler(rewritten, rctx.headers()) {
                        @Override
                        protected byte[] onSendRequestMessage(byte[] data, boolean binary) {
                            wsRequestMessages.incrementAndGet();
                            return data;
                        }

                        @Override
                        protected byte[] onSendResponseMessage(byte[] data, boolean binary) {
                            wsResponseMessages.incrementAndGet();
                            return data;
                        }
                    };
                }
            };

            try (CopilotClient client = newLlmClient(ctx, handler,
                    "COPILOT_EXP_COPILOT_CLI_WEBSOCKET_RESPONSES=true")) {
                CopilotSession session = client
                        .createSession(new SessionConfig().setOnPermissionRequest(PermissionHandler.APPROVE_ALL)).get();

                AssistantMessageEvent result = session.sendAndWait(new MessageOptions().setPrompt("Say OK.")).get(60,
                        TimeUnit.SECONDS);
                session.close();

                // The HTTP seam fired — the runtime issued model-layer GETs (catalog,
                // policy) and possibly a single-shot inference through the send override.
                assertTrue(httpRequests.get() > 0, "Expected the HTTP send override to fire");
                assertTrue(httpResponses.get() > 0, "Expected the HTTP response mutation to fire");

                // The WebSocket seam fired — the main agent turn went over the WS path and
                // we observed messages in both directions.
                assertTrue(wsRequestMessages.get() > 0, "Expected runtime -> upstream ws messages");
                assertTrue(wsResponseMessages.get() > 0, "Expected upstream -> runtime ws messages");
                assertTrue(upstream.upstreamWsRequests() > 0, "Expected the upstream WS to receive request messages");

                // Validate the final assistant response arrived (guards against truncated
                // captures)
                String text = assistantText(result);
                assertTrue(text.contains("OK from synthetic") && text.contains("upstream"),
                        "Expected synthetic upstream content in assistant reply, got " + text);
            }
        }
    }
}
